### PR TITLE
Make derivation of total column ozone (`toz`) more flexible and add derivation of stratospheric and tropospheric column ozone

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR makes the derivation of total column ozone (`toz`) more flexible by allowing its derivation from
- data that has hybrid levels (e.g., `o3` from the AERmon table)
- zonal mean data (e.g., `o3` from the AERmonZ table)

In addition, this PR adds two derived variables:
- stratospheric column ozone (`soz`), where the "stratosphere" is defined as the region where the O3 mole fraction exceeds 125 ppb
- tropospheric column ozone (`troz`), where the "stratosphere" is defined as the region where the O3 mole fraction is lower than 125 ppb

### Backwards-incompatible change

This PR also changes the units of `toz` in our custom table from DU to m. This is done to make it consistent with the [CMIP6 definition](https://github.com/ESMValGroup/ESMValCore/blob/4c21ca64185cc253ac896ee13c623afd168d7bc3/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_AERmon.json#L1979). Otherwise, this will lead to various problems when directly comparing CMIP6 output with CMIP5 output (or another project that uses the custom table).

To restore the old behavior, a [preprocessor](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/recipe/preprocessor.html#convert-units) can be added in the recipe to convert m to DU:

```yml
convert_units:
  units: DU
```

***

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] ~Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)~ (see above)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
This can be tested with the following recipe:

```yml
# ESMValTool
---
documentation:
  title: test
  description: test
  authors:
    - schlund_manuel
  maintainer:
    - schlund_manuel

preprocessors:
  du:
    convert_units:
      units: DU

diagnostics:

  test:
    variables:
      toz:
        derive: true
        force_derivation: true
        exp: historical
        start_year: 2013
        end_year: 2014
        preprocessor: du
        additional_datasets:
          - {project: CMIP6, dataset: CESM2-WACCM, ensemble: 'r1i1p1f1', grid: gn, mip: Amon}
          - {project: CMIP6, dataset: GISS-E2-1-G, ensemble: 'r1i1p3f1', grid: gn, mip: AERmon}
      troz:
        derive: true
        force_derivation: true
        exp: historical
        start_year: 2013
        end_year: 2014
        preprocessor: du
        additional_datasets:
          - {project: CMIP6, dataset: CESM2-WACCM, ensemble: 'r1i1p1f1', grid: gn, mip: Amon}
          - {project: CMIP6, dataset: GISS-E2-1-G, ensemble: 'r1i1p3f1', grid: gn, mip: AERmon}
      soz:
        derive: true
        force_derivation: true
        exp: historical
        start_year: 2013
        end_year: 2014
        preprocessor: du
        additional_datasets:
          - {project: CMIP6, dataset: CESM2-WACCM, ensemble: 'r1i1p1f1', grid: gn, mip: Amon}
          - {project: CMIP6, dataset: GISS-E2-1-G, ensemble: 'r1i1p3f1', grid: gn, mip: AERmon}
    scripts:
      null
```

[recipe_test_ozone.yml.txt](https://github.com/user-attachments/files/16601226/recipe_test_ozone.yml.txt)

**Comment by schlunma** [comment]
As far as I can tell with, toz and soz look good. However, I get some ugly results for troz, which include lots of artifacts:

grafik

Most likely this is the result of lower pressure levels being masked in the input data (o3):

grafik

**Comment by valeriupredoi** [comment]
Manu, can you name these islands?
islands

**Comment by schlunma** [comment]
> Manu, can you name these islands? ![islands](https://private-user-images.githubusercontent.com/28983971

Not all of them, but Google Maps certainly can 😄

**Comment by FranziskaWinterstein** [comment]
Hey @schlunma, thank you for taking care of this. 

I made my own attempt to build troz.py (see below), by basically calculating troz as troz = toz - soz. This reduces the doubling of code, but introduces dependencies from troz on soz. What do you think about that?

```
"""Derivation of variable ``soz``."""

import dask.array as da
import iris

from ._baseclass import DerivedVariableBase
from .toz import DerivedVariable as Toz
from .soz import DerivedVariable as Soz

class DerivedVariable(DerivedVariableBase):
    """Derivation of variable ``troz``."""

    @staticmethod
    def required(project):
        """Declare the variables needed for derivation."""
        if project == 'CMIP6':
            required = [{'short_name': 'o3'}, {'short_name': 'ps'}]
        else:
            required = [{'short_name': 'tro3'}, {'short_name': 'ps'}]
        return required

    @staticmethod
    def calculate(cubes):
        """Compute tropspheric column ozone.

        Note
        ----
        In the calculation of ``troz``, the surface air pressure (``ps``) is
        used to determine the pressure level width of the lowest layer.

        The calculation of ``troz`` consists of three steps:
        (1) Use derivation function of ``toz`` to calculate ``toz``.
        (2) Use derivation function of ``soz`` to calculate ``soz`` (using the
            masked data).
        (3) Calculate difference of ``toz`` and ``soz`` to calculate ``troz``.

        """
        o3_cube = cubes.extract_cube(
            iris.Constraint(name='mole_fraction_of_ozone_in_air'))
        ps_cube = cubes.extract_cube(
            iris.Constraint(name='surface_air_pressure'))

        # (1) Use derivation function of toz to calculate toz
        cubes = iris.cube.CubeList([o3_cube, ps_cube])
        cube_toz = Toz.calculate(cubes)

        # (2) Use derivation function of soz to calculate soz
        cube_soz = Soz.calculate(cubes)

        # (3) Calculate troz as difference of toz and soz
        return cube_toz - cube_soz
```

**Comment by FranziskaWinterstein** [comment]
I successfully tested the given recipe as well as applied it to the diagnostics to plot `hovmoeller_vs_lat_or_lon` and `map` plots from multi_dataset.py. No problems.

**Comment by FranziskaWinterstein** [comment]
The map plots of the climatology 2013 - 2014 of `troz` do not show those artefacts. I suppose they only appear without time averaging. If it is data dependent, I guess, there is not much we can do about it?
map_CESM2-WACCM_troz
Created with:
[recipe_test_ozone_map.yml.txt](https://github.com/user-attachments/files/16896914/recipe_test_ozone_map.yml.txt)

File: esmvalcore/preprocessor/_derive/_shared.py [hunk 1]
```diff
@@ -170,15 +170,15 @@ def _create_pressure_array(cube, ps_cube, top_limit):
 def _get_pressure_level_widths(array, air_pressure_axis=1):
     """Compute pressure level widths.
 
-    For a 1D array with pressure level columns, return a 1D array with
+    For a 4D array with pressure level columns, return a 4D array with
```
Comment (FranziskaWinterstein) [inline esmvalcore/preprocessor/_derive/_shared.py:173]: Is 4D mandatory or does it work for other dimensions (given that air_pressure is available)?

**Review by FranziskaWinterstein — APPROVED** [review]
Thank you for sharing this optional of ozone variables. I am impressed by the comprehensive testing.

File: esmvalcore/preprocessor/_derive/_shared.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_derive/_shared.py:173]: You're right, this should work for any array

**Comment by schlunma** [comment]
> I made my own attempt to build troz.py (see below), by basically calculating troz as troz = toz - soz. This reduces the doubling of code, but introduces dependencies from troz on soz. What do you think about that?

I first had it that way and it produced exactly the same results. However, you trade simpler code for speed in this case, as this would mean that the (potential) interpolation needs to be done twice, which can be very expensive. I think avoiding this is better than saving ~20 lines of code.

> The map plots of the climatology 2013 - 2014 of troz do not show those artefacts. I suppose they only appear without time averaging. If it is data dependent, I guess, there is not much we can do about it?

Ah, nice to see that time averaging helps! I also think that there's nothing we can really do about it...

Thanks for reviewing 🎉

**Comment by schlunma** [comment]
@valeriupredoi since you already commented on this PR, would you do us the honor of performing a technical review? Thanks 🍻

**Comment by valeriupredoi** [comment]
> @valeriupredoi since you already commented on this PR, would you do us the honor of performing a technical review? Thanks 🍻

on it right now, bud :beer:

File: esmvalcore/preprocessor/_derive/toz.py [hunk 2]
```diff
@@ -19,16 +22,88 @@
...
+        15000.0,
+        10000.0,
+        7000.0,
+        5000.0,
+        3000.0,
+        2000.0,
+        1000.0,
+        500.0,
+        100.0,
+    ]
```
Comment (valeriupredoi) [inline esmvalcore/preprocessor/_derive/toz.py:82]: these should really be loaded from the CMIP6_coords CMOR table, not hardwired here

**Review by valeriupredoi — APPROVED** [review]
this looks very good, many thanks! My only (more) serious concern is with those plev points that should be loaded straight from the CMOR table, but all else looks good, cheers :beer:

File: esmvalcore/preprocessor/_derive/soz.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                   o3_cube.coord_dims('latitude') +
+                   o3_cube.coord_dims('longitude'))
+        idx_to_extract_ps = [0] * o3_cube.ndim
+        for ps_dim in ps_dims:
+            idx_to_extract_ps[ps_dim] = slice(None)
+        ps_cube = o3_cube[tuple(idx_to_extract_ps)].copy()
+
+        # Set ps data using lowest pressure level available and add correct
+        # metadata
+        lowest_plev = o3_cube.coord('air_pressure').points.max()
```
[the remark this answers is not in the document]
Comment (schlunma) [reply esmvalcore/preprocessor/_derive/soz.py:89]: Yeah, "low" means low in the atmosphere...can't think of a better name tbh 😬

File: esmvalcore/preprocessor/_derive/toz.py (hunk 2 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_derive/toz.py]: Done in https://github.com/ESMValGroup/ESMValCore/pull/2509/commits/c344b95f408bae7f6486fd164088319feaf923f8

File: esmvalcore/preprocessor/_derive/toz.py [hunk 4]
```diff
@@ -19,16 +22,88 @@
...
+def ensure_correct_lon(o3_cube, ps_cube=None):
+    """Ensure that ``o3`` cube contains ``longitude`` and adapt ``ps`` cube."""
+    if o3_cube.coords('longitude'):
+        return (o3_cube, ps_cube)
+
+    # Get zonal mean ps if necessary
+    if ps_cube is not None:
+        ps_cube = ps_cube.collapsed('longitude', iris.analysis.MEAN)
+        ps_cube.remove_coord('longitude')
+
```
[the remark this answers is not in the document]
Comment (schlunma) [reply esmvalcore/preprocessor/_derive/toz.py]: Good point! I tried to do that in https://github.com/ESMValGroup/ESMValCore/pull/2509/commits/20751276ec0adee0d6a38e6fa6f48a5bd700317f (and also fixed a tiny bug in the troz derivation). Please merge if you are happy with this 👍

[the remark this answers is not in the document]
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_derive/toz.py]: Cheers, bud! Lemme have looksee then will merge

File: esmvalcore/preprocessor/_derive/toz.py (hunk 2 above)
Comment (valeriupredoi) [reply esmvalcore/preprocessor/_derive/toz.py]: Outstanding, private first class 😁

**Pull request merged**